### PR TITLE
Epic 7.7.1: Syllabus Generator - Logic Legion

### PR DESCRIPTION
**Pull request by chitoroagad**
# Syllabus Generator

## Summary:
- This tool aims to create a sample syllabus customised to the liking of the user.
- Users can specify a small number of specific inputs as well as vague instructions to the tool and receive a tailored syllabus.
- Users can specify which sections of the syllabus to include or exclude in their output using the `options` api parameter.
- The generated syllabus includes the following:
    - Title
    - Overview
    - Policies and Exceptions
    - Grading Policy
    - Assessment Components
    - Required/Recommend Materials 

## Changes:
1. syllabus_generator directory in `features/` containing:
    - `tools.py` includes `SyllabusBuilder` class responsible for all logic involved in using the LLM to create the syllabus.
    - `model.py` includes the Pydantic model for validating output.
    - `core.py` includes the executor function.
    - `metadata.json` includes expected inputs for the api.
    - `prompt/` includes various text files for prompt generation.
    - `tests/` includes `test_sb.py` used for testing all components of the SyllabusBuilder as well as `test.json` which includes a sample input for the api.
2. Updated the `tools_config.json` to be able to access the syllabus_generator straight from the api.
3. Created development environment in `flake.nix` for better LSP support and simplified environment set up.
    - Activate using `nix develop` and `--command <prefered shell>` to enter env.
    - `bash` is default.

### Input
```json
    {
        "user": {
            "id": "string",
            "fullName": "string",
            "email": "string"
        },
        "type": "tool",
        "tool_data": {
            "tool_id": 2,
            "inputs": [
                {
                    "name": "subject",
                    "value": "Data Structures"
                },
                {
                    "name": "grade_level",
                    "value": "University"
                },
                {
                    "name": "course_overview",
                    "value": "This course covers the fundamental concepts and applications of data structures in computer science. Students will explore various data structures such as arrays, linked lists, stacks, queues, trees, and graphs. The course will emphasize both the theoretical and practical aspects of data structures, including their implementation and analysis of their efficiency. By the end of the course, students will be equipped with the knowledge to use data structures effectively in real-world applications and advanced computing challenges."
                },
                {
                    "name": "customisation",
                    "value": "I want the course to require proficiency in Python"
                },
                {
                    "name": "options",
                    "value": [
                        "all"
                    ]
                }
            ]
        }
    }
```

### Output
```json
{
  "data": {
    "title": "Data Structures",
    "overview": "This course covers the fundamental concepts and applications of data structures in computer science. Students will explore various data structures such as arrays, linked lists, stacks, queues, trees, and graphs. The course will emphasize both the theoretical and practical aspects of data structures, including their implementation and analysis of their efficiency. By the end of the course, students will be equipped with the knowledge to use data structures effectively in real-world applications and advanced computing challenges.",
    "objectives": [
      "Understand the fundamental concepts of data structures, including their definition, representation, and operations.",
      "Apply data structures to solve real-world problems in various domains.",
      "Analyze the efficiency of data structures in terms of time and space complexity.",
      "Implement data structures in Python and evaluate their performance.",
      "Develop a deep understanding of the strengths and weaknesses of different data structures."
    ],
    "policies_and_exceptions": {
      "attendance_requirements": "Regular attendance is expected and absences must be communicated in advance. More than two unexcused absences may result in a deduction of your final grade.",
      "make-up_work": "Missed assignments and exams must be made up within one week of the original due date. Make-up exams may be different from the original exam.",
      "academic_honesty": "All work submitted for this course must be original and your own. Any instances of plagiarism will result in severe penalties, including failing the course."
    },
    "grade_level_assessments": {
      "assessment_components": {
        "assignments": 20,
        "exams": 30,
        "projects": 30,
        "participation": 10,
        "quizzes": 10
      },
      "grade_scale": {
        "A": "90-100%",
        "B": "80-89%",
        "C": "70-79%",
        "D": "60-69%",
        "F": "Below 60%"
      },
      "performance_expectations": "Students are expected to demonstrate a high level of understanding of the course material, participate actively in class discussions, and complete all assignments and projects on time.",
      "feedback_and_improvement": "Feedback on assignments and exams will be provided within one week of submission. Students are encouraged to seek additional feedback during office hours or by appointment."
    },
    "required_materials": {
      "recommended_books": [
        "Data Structures and Algorithms in Python",
        "Data Structures and Algorithms Made Easy",
        "Introduction to Algorithms"
      ],
      "required_items": [
        "Laptop with Python installed",
        "Notebook and pens",
        "Access to online learning platform"
      ]
    },
    "additional_information": {
      "prerequisites": "Basic programming skills in Python",
      "office_hours": "Mondays and Wednesdays, 2-3 PM",
      "contact_information": "Email: username@example.edu, Phone: 123-456-7890"
    }
  }
}
```

### Input Validation Tests:

• `test_invalid_input`: Ensures that invalid inputs such as an empty subject or
missing grade level raise appropriate errors.

• `test_valid_model` and `test_invalid_model`: Validates the Pydantic model
(SyllabusModel) to ensure that it correctly accepts valid data and raises
errors for missing or incorrect fields.

### Functionality Tests:

• `test_create_prompt_temp_with_mocked_read_text_file`: Checks if the
`create_prompt_temp` function correctly assigns the appropriate grading
policy based on the grade level, using mock inputs to simulate different
school levels.

• `test_compile`: Verifies that the compile function correctly generates the
prompt and model chain based on the type of compilation ("customization"
or "syllabus").

• `test_validate_response_valid` and `test_validate_response_invalid`: Checks
the `validate_response` method to ensure it correctly identifies valid and
invalid syllabus structures.

## How Tests Were Conducted

The tests were conducted using the pytest framework, which provides a
robust environment for running test cases. This includes using fixtures for
setting up test cases (sb fixture for the SyllabusBuilder instance) and
mocking functions to simulate and control the behavior of dependencies.

## Cases Covered

1. Input validation to ensure that the correct types and values are passed.

2. Prompt creation based on different grade levels to ensure the proper
   grading policies are applied.
3. Syllabus compilation to ensure that the model and prompt chains are
   correctly assembled and invoked.
4. Model validation to confirm that the syllabus structure adheres to the
   expected schema.
5. Response validation to make sure that the generated syllabus meets all
   necessary requirements.

### Input Validation Tests

`test_invalid_input`:

```python
with pytest.raises(ValueError):
    SyllabusBuilder(subject='', grade_level=10)

with pytest.raises(ValueError):
    SyllabusBuilder(subject='Biology', grade_level=None)
```

`test_valid_model` and `test_invalid_model`:

```python
valid_input = {
    'title': 'Sample Syllabus Title',
    'overview': 'Sample overview of the syllabus content.',
    'objectives': [
        'Define the key terms associated with...',
        'Describe the cardiac cycle...',
    ],
    'policies_and_exceptions': {
        'attendance_requirements': 'Sample attendance requirements...',
        'make_up_work': 'Sample make-up work policy...',
    },
    'grade_level_assessments': {
        'assessment_components': {
            'assignments': 20,
            'exams': 25,
            'projects': 25,
            'presentations': 15,
            'participation': 15,
        },
        'grade_scale': {
            'A': '90-100%',
            'B': '80-89%',
            'C': '70-79%',
            'D': '60-69%',
            'F': 'Below 60%',
        },
    },
    'required_materials': {
        'recommended_books': ['book 1', 'book 2', 'book 3'],
        'required_items': [
            'paint',
            'paintbrush',
            'pencil',
            'eraser',
            'notebooks', 'sharpies',
            'crayons',
            'black ink pen',
            'ruler',
        ],
    },
}
```

### Functionality Tests

`test_compile`:

```python
@patch.object(SyllabusBuilder, '\_create_custom_promptTemp')
@patch.object(SyllabusBuilder, '\_create_prompt_temp')
def test_compile(mock_create_prompt_temp, mock_create_custom_promptTemp):
    mock_custom_prompt_instance = MagicMock(spec=PromptTemplate)
    mock_create_custom_promptTemp.return_value = mock_custom_prompt_instance

    mock_syllabus_prompt_instance = MagicMock(spec=PromptTemplate)
    mock_create_prompt_temp.return_value = mock_syllabus_prompt_instance

    sb = SyllabusBuilder(subject='Mathematics', grade_level='Grade 5', verbose=True)

    chain_customisation = sb._compile('customisation')
    mock_create_custom_promptTemp.assert_called_once()
    chain_syllabus = sb._compile('syllabus')

    mock_create_prompt_temp.assert_called_once()
```
Testing documentation by @ElyasBinothman 

## Documentation made by team members
[Syllabus Generator.pdf](https://github.com/user-attachments/files/16564543/Syllabus.Generator.pdf)  @YomnaEisa

### CODE REVIEW

File: .gitignore
Comment (mikhailocampo) [inline .gitignore:1]: Update the PDF to include a summary of what the tests were and what edge cases were accounted for

File: app/features/syllabus_generator/core.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    except Exception as e:
+        error_message = f"Error in executor: {e}"
+        logger.error(error_message)
+        raise ValueError(error_message)
+
+    return syllabus
+
+
+if __name__ == "__main__":
```
Comment (mikhailocampo) [inline app/features/syllabus_generator/core.py:59]: Remove debugging section

File: app/features/syllabus_generator/tools.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+    # Probably a better way to do this
+    # TODO: Make this return ENUM so we can attach which part is broken
+    def _validate_response(self, response: Dict) -> bool:
+        """
+        Validates response from LLM
+        """
+        try:
+            # Assuming reponse is already a dict
+            if not isinstance(response, dict):
```
Comment (mikhailocampo) [inline app/features/syllabus_generator/tools.py:111]:
```
from pydantic import BaseModel, Field, model_validator
from typing import Optional
from typing_extensions import Annotation

class LLMOutput(BaseModel):
  title: str
  overview: str
  description: Annotation[str, Field(default="This is a default desc.")]
  assignments: Annotation[List[Assignment], Field(default_factory=list)] 
  ....
```

Pydantic for validation can be useful here for validating the output

File: app/features/syllabus_generator/tools.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+                if "policies_and_exceptions" in response:
+                    policies_and_exceptions = response["policies_and_exceptions"]
+                    if not isinstance(policies_and_exceptions, dict):
+                        print(5)
+                        return False
+                    for key, value in policies_and_exceptions.items():
+                        if not isinstance(key, str) or not isinstance(value, str):
+                            print(6)
+                            return False
+
```
Comment (mikhailocampo) [inline app/features/syllabus_generator/tools.py:143]: If you need to validate response fields with Pydantic's `model_validator`

```python
from pydantic import BaseModel, model_validator, Field

class MultipleChoiceQuestion(BaseModel):
  answer: str
  choices: List[str]
  
  @model_validator()
  def validate_answer(cls, v):
    # Checks to make sure answer exists in the choice 
    assert answer in choice
```

File: flake.lock
Comment (mikhailocampo) [inline flake.lock:1]: Not sure if this is needed to be committed

**Review by mikhailocampo — CHANGES_REQUESTED** [review]
Please see changes

File: flake.lock
Comment (chitoroagad) [reply flake.lock:1]: This makes sure everyone is using the same versions of packages.
Can be updated using `nix flake update`.

The whole flake.nix is more of a quality of life change for those familiar with nix.

File: app/features/syllabus_generator/tools.py (hunk 2 above)
Comment (chitoroagad) [reply app/features/syllabus_generator/tools.py:111]: I couldn't find any documentation on `Annotation` but I found `Annotated`, I assume this is what you meant.
Should I use this over just `Field`  in `model.py`.

File: app/features/syllabus_generator/core.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+from typing import List
+from services.logger import setup_logger
+from app.features.syllabus_generator.tools import SyllabusBuilder
+from app.api.error_utilities import LoaderError, ToolExecutorError
+
+logger = setup_logger()
+
+
+def executor(
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/core.py:9]: Use a Pydantic schema for decoupling the args.

File: app/features/syllabus_generator/core.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    **kwargs,
+):
+    """Execute the syllabus generation process."""
+    try:
+        if verbose:
+            logger.debug(
+                f"Subject: {subject}, Grade Level: {grade_level}, Course Overview: {course_overview}"
+            )
+
+        sb = SyllabusBuilder(
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/core.py:25]: Use a Pydantic schema for decoupling the args.

File: app/features/syllabus_generator/schema.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+
+class SyllabusModel(BaseModel):
+    title: Optional[str] = Field(description="The title for the whole course")
+    overview: Optional[str] = Field(
+        description="A broad overview of what is expected of students and what they will learn"
+    )
+    objectives: Optional[List[str]] = Field(
+        description="A list of specific tasks the student will be able to successfully do upon completion of the course",
+        examples=[
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/schema.py:13]: Good way for enabling deterministic outputs.

File: app/features/syllabus_generator/tools.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    "overview": "Generate a comprehensive overview of the course.",
+    "objectives": "Generate a list of learning objectives to show a student's understanding.",
+    "required_materials": "Generate a list of materials that are required by the students to successfully participate in the course. Materials could be things like books, tools or supplies. Use two key variables in the output of required_materials. One is recommended_books and another is required_items.",
+    "additional_information": "Include any additional information that could be useful to the student",
+    "policies_and_exceptions": "Create a dictionary of policies and exceptions that apply to the course.",
+    "grade_level_assessments": "{grade_level_assessments}",
+}
+
+# Options that can be added to prompt
+USABLE_OPTIONS = ALLOWED_OPTIONS.copy()
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:34]: Why is this shallow copy needed?

File: app/features/syllabus_generator/tools.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+USABLE_OPTIONS.pop("all", None)
+
+
+class SyllabusBuilder:
+    def __init__(
+        self,
+        subject: str,
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:41]: Use a Pydantic schema for decoupling args.

File: app/features/syllabus_generator/tools.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        customisation: str = "",
+        options: List[str] = ["all"],
+        prompt: str = "",
+        model=None,
+        parser=None,
+        verbose: bool = False,
+    ):
+        """Initialize SyllabusBuilder with default configurations."""
+        default_config = {
+            "model": GoogleGenerativeAI(model="gemini-1.0-pro"),
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:53]: Replace this model to `gemini-1.5-flash` or `gemini-1.5-pro` as that model is deprecated.

File: app/features/syllabus_generator/tools.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+        }
+
+        self._validate_inputs(subject, grade_level, options, customisation)
+
+        self.prompt = prompt or default_config["prompt"]
+        self.model = model or default_config["model"]
+        self.parser = parser or default_config["parser"]
+        self.grade_level_assessments = ""
+        self.customisation = customisation
+        self.options = list(map(lambda x: x.strip().lower(), options))
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:65]: Good way of setting the options.

File: app/features/syllabus_generator/tools.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+        self.subject = subject
+        self.grade_level = grade_level.strip().lower()
+        self.course_overview = course_overview
+        self.verbose = verbose
+
+        self._post_validate_inputs()
+
+    def _validate_inputs(self, subject, grade_level, options, customisation):
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:74]: Good way of differentiating between public and private methods

File: app/features/syllabus_generator/tools.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+                "format_instructions": self.parser.get_format_instructions()
+            },
+        )
+        return prompt
+
+    # Returns langchain chain for creating syllabus
+    def _compile(self, case: str):
+        if case == "syllabus":
+            prompt = self._create_prompt_temp()
+        elif case == "customisation":
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:181]: Why is the `customization` part needed?

File: flake.nix
Comment (AaronSosaRamos) [inline flake.nix:1]: This file is not needed as it's not a part of the main project. Please, remove it.

File: flake.lock
Comment (AaronSosaRamos) [inline flake.lock:1]: This file is not needed as it's not a part of the main project. Please, remove it.

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Please, solve the given comments.

**Comment by AaronSosaRamos** [comment]
Closed due inactivity
